### ISSUE TITLE
Consider keybinding platform context keys on start up

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#85058]
I have these keybindings

```
{ "key": "ctrl+n",           "command": "explorer.newFile", "when": "explorerViewletVisible && filesExplorerFocus && !isMac" },
{ "key": "cmd+n",            "command": "explorer.newFile", "when": "explorerViewletVisible && filesExplorerFocus && isMac" },
{ "key": "ctrl+shift+n",     "command": "explorer.newFolder", "when": "explorerViewletVisible && filesExplorerFocus && !isMac" },
{ "key": "cmd+shift+n",      "command": "explorer.newFolder", "when": "explorerViewletVisible && filesExplorerFocus && isMac" },

{ "key": "cmd+q",            "command": "-workbench.action.quit", "when": "isMac" },
{ "key": "cmd+q cmd+q",      "command": "workbench.action.quit", "when": "isMac" }
```

Which end up showing the keybindings in menus since contexts are not considered here.

![image](https://user-images.githubusercontent.com/2193314/69081275-76e9f080-09f2-11ea-9084-1e8fb60eb6d3.png)
![image](https://user-images.githubusercontent.com/2193314/69081295-81a48580-09f2-11ea-9ddb-2b6419aba38b.png)

Maybe we could refine the `when` clause just for startup to ignore all entries except for `isMac`, `isWindows` and `isLinux`? For example:

```
explorerViewletVisible && filesExplorerFocus && !isMac
= (true) && (true) && !isMac
= !isMac
```

For or queries we can just assume everything is true?

```
terminalFocus && isWindows || terminalFocus && isLinux || filesExplorerFocus
= (true) && isWindows || (true) && isLinux || (true)
= isWindows || isLinux || true
= true
```

**Comment by sbatten** [comment]
/cc @alexandrudima 

I implemented what this could be like yesterday. I updated contextkeys to use an optional `includeKeys` when evaluating and treat everything else as true. To make this cleaner, the keybinding service and/or resolver would probably surface a flag or something like `platformsOnly` or `staticOnly`.

It feels hacky though so not sure this is a good direction as far as implementation. The idea sounds reasonable to me though.

**Comment by alexandrudima** [comment]
@sbatten IMHO I think evaluation is "too late".

I would tackle this specific issue in the following way:
- introduce the concept of a constant context key registry.
- e.g. on macs, we can register the following constants: `isMac: true`, `isWindows: false`, `isLinux: false`
- then, add support for this constant context key registry in all the `ContextKey*Expr.create` methods.
- e.g. calling `ContextKeyAndExpr.create(ContextKeyDefinedExpr.create('a'), ContextKeyDefinedExpr.create('isMac'))` on a mac would result in `ContextKeyDefinedExpr.create('a')`. i.e. `create('a && isMac')` ==> `create('a')`.

In other words, I would directly eliminate the OS keys at parse time.

Note: To really work well, we might need to introduce two new classes `ContextKeyTrueExpr` and `ContextKeyFalseExpr` and add support for those in the keybindings rules, so we can completely discard keybindings which will never hold as soon as we parse the `keybindings.json` file.

**Comment by Tyriar** [comment]
Thanks a lot for this, works great with my setup!

![image](https://user-images.githubusercontent.com/2193314/75687600-e22e9a00-5c52-11ea-950e-2436be45327c.png)
![image](https://user-images.githubusercontent.com/2193314/75687639-eeb2f280-5c52-11ea-933e-8890f574bbf0.png)
